### ISSUE TITLE
API v1 discussion

### ISSUE THREAD

**Issue by povrazor** [ludumdare/ludumdare#1175]
Here's a central issues to discuss the API changes that need to happen before we publish API v1.

**Comment by sgstair** [comment]
Some rough thoughts on API changes:

1. node and note love requests should be POST (They have side effects)
2. I think it might make sense to fragment off love/star and other kinds of markers into a separate endpoint, something like /mark/star/add/node/<id>, /mark/love/remove/note/<id>
3. I want to be able to fetch a feed that mixes nodes and notes (e.g. get my recent posts and comments)
4. Need better querying for individual comments, get user comments, stuff like this.
5. We talked some about unifying feed into a new endpoint with common code to deal with feed options, and then minimal providers to generate feeds for different object types.

**Comment by povrazor** [comment]
# General
I would like to see code like this

```
		$Node.Get(node.parent)
		.then(r => {
			if ( r.node && r.node.length ) {
				this.setState({'parent': r.node[0]});
				
				return $Grade.GetMy(node.id);
			}
			return Promise.resolve({});
		})
```
specifically this part:
```
			if ( r.node && r.node.length ) {
```
streamlined a bit.

What I mean: the raw responses from the API are being passed to the requests. The code could be changed to:

```
			if ( r && r.length ) {
```

or

```
			if ( r && !r.error ) {
```

Many API requests are allowed to fail (no items in feed), and it needs to be decided where true errors (catch) should be used.

Plus, we have very-basic async/await support now in the compiler, so this main block of code above could potentially be simplified.

# Feeds
Feeds return arrays of nodes that should be displayed in an order. Such as blog feeds, game lists, results, etc.

* Feeds are complex. They should definitely become their own file

# Walk
Walk is used to decypher a URL request by stepping through each element of the path, to find the actual node. The remaining nodes are passed to the node type as extra arguments.

* Walk needs to return a timestamp, so it too can request reloading of data upon change (like feeds)
* Walk *could* include node data as an optimization, since a walk always returns 1 node

**Comment by local-minimum** [comment]
On the __general__, it sounds like me that `if (r.valid)` would be even nicer, but I'm also wondering here, if the preprocessing shouldn't repackage the response so that you could have:

```js
$Node.Get(node.parent)
  .then(
    r => {
      this.setState({'parent': r.node[0]});
      return $Grade.GetMy(node.id);
    },
    r => {
      //Handle all sorts of issues if this is how the promises look like
    });
```

That is preprocessing should reject all things that could be issues and you should not have to test validity in the successful part of `then`.

I never really understood how __walk__ works. Still don't. Which should be an indication of that it would be nice if the mechanism could be simplified or clarified somehow. 

I do agree (if I understand you correctly) that __feeds__ are quite distinct from common nodes and would be better treated as entirely separate entities.

**Comment by povrazor** [comment]
There are cases where returning an empty array is valid, and I suspect the same goes for an empty response, but I don't yet know. At the moment error handling tends to get collapsed in to the success case. It begs the question where errors should be raised. 

Raising errors to the user should probably be revisited too. There was some hackery in place for errors related to signups, but there should be a more standard way to inform of errors.

I'm also more leaning towards `.catch(...)` instead of the 2nd argument to `.then(...)`, just because it's more clear that it's an error because of the word `catch`.

**Comment by sgstair** [comment]
Any APIs that have side effects should really be POST rather than GET (I found a bunch more in grading, theme voting) - While this isn't a huge functional issue because over HTTPS it's unlikely for a caching proxy to rely on the meaning of GET vs POST semantics, there may still be situations where this causes hiccups.

**Comment by povrazor** [comment]
I'm not expecting any hiccups, but I'm with you for the consistency.

**Comment by local-minimum** [comment]
I'm probably using `$Node.Get` the wrong way, but if I have an array of nodes I want to get, I manually join their id on '+'. Most if not all of these individual nodes actually exists in the cache, but since I haven't done the exact same request before, I don't use the cache. Or I think this is happening. This seems like a waste.

**Comment by povrazor** [comment]
You should be able to pass the array directly to `$Node.Get`. Get should support both standard array-of-ids format, and array-of-ids-with-changes format.

**Comment by local-minimum** [comment]
Yeah that was it, Sets were not allowed, which is a bit of a pitty because they are nice things for making sure you don't ask for the same node more than once.

**Comment by sgstair** [comment]
Note support for Set is a bit of a [minefield](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#Browser_compatibility) in IE (only partial support), so we may need to add more compat shims. Not sure if you have tested in IE :)

**Comment by local-minimum** [comment]
No I didn't but I assumed our ES6 compiler would take care of these as well. (An my general feeling about ppl using IE is that they don't deserve support, they should at least have switched to Edge).

**Comment by povrazor** [comment]
The compiler more handles syntax than features. That said we do have polyfills in place to add support for features not found in older browsers, but I'm not familiar enough with the problem with sets to know if we have that covered or not. 

As far as legacy browser compatibility goes, I want the site to work on IE11 (i.e. Windows 7), but that's it. Windows 7 users don't have the luxury to install Edge. You need to sacrifice your PC to the garbage that is Windows 8 and 10 to get Edge.

**Comment by povrazor** [comment]
Consider changing all API's that specify an `ID` as part of the path to include the `$`.

eg.

`https://api.ldjam.com/vx/feed/$1/parent`

This is the syntax used by other URL decoders to differentiate IDs from slugs. It may as well be standard.

**Comment by AndreasHae** [comment]
What's the state of this issue? What still needs to be done to publish stable API v1?

**Comment by mikekasprzak** [comment]
Off the top of my head there's a bunch of unfinished code that needs to be merged in (the new `/feed`, `/search`, and improved `/node/meta` APIs). Maybe some initial-batch APIs too. Then do a once-over of everything to make sure everything is being done properly (POST instead of GET in the cases that @sgstair pointed out).
